Added "safe delete" feature

rafaelsierra opened the pull request:
When enabled, calling `.delete()` on a file will cause it to be moved
somewhere else to be retrieved back if necessary.

It is possible to use `strftime` formatting on the backup object name

Cleaning up the backup folder is not included

This feature came up so we could fulfill contract requirements with backup clauses. With this feature we can backup minio using `mc mirror --remove minio/ /backup/folder/` and still delete files normally.

Commit:
Added "safe delete" feature

When enabled, calling `.delete()` on a file will cause it to be moved
somewhere else to be retrieved back if necessary.

It is possible to use `strftime` formatting on the backup object name

Cleaning up the backup folder is not included

thomasf commented:
I'm thinking that we maybe could come up with a better name,, safe_delete signals that deletion will happen when it don't.. Maybe even a more explicit `rename_on_delete` could work even i it's a bit verbose?

thomasf commented:
Great feature though, thanks for submitting.

rafaelsierra commented:
I thought the same about naming it as "safe delete", since it can also be interpreted as "hey, your file will be super deleted, there is no way to recover it back" (like the `srm` tool).

Maybe `backup_on_delete`? Since you can move to another bucket and not just rename it

thomasf commented:
> Maybe backup_on_delete? Since you can move to another bucket and not just rename it

yes, let's go with that.

thomasf commented:
`MINIO_STORAGE_SAFE_DELETE` can probably be removed and the feature activated when the `MINIO_STORAGE_SAFE_DELETE_PATH` is used..  Less settings are usually good to avoid confusing settings files.

`MINIO_STORAGE_SAFE_DELETE_PATH` could maybe be named  something like `MINIO_STORAGE_MEDIA_BACKUP_FORMAT` to be clear about that it is a format string and not a static path.

Maybe also add `MINIO_STORAGE_STATIC_BACKUP_FORMAT` unless we just skip to support this on staticfiles storage.. I can't see why anyone would want backup on static files. If someone really needs it they can subclass their own storage in their application.

Commit:
Renamed feature from "safe delete" to "backup on delete"

rafaelsierra commented:
I am avoiding distinguishing between media and static files because we use this module as neither, we are using directly as a storage backend for `FileField`s, in other words, we are not using Django's media system.

I like the idea to remove the flag settings and rely on the path, but that raises the question of what to do when the bucket is not set, should we raise an error or default to something else?

I agree that "path" might be misleading and "format" makes it clear that some processing will happen on the string

Do you still think that media and static files should be treated differently (i.e. by introducing more settings) or for simplicity sake let the backup work on static files as well? I totally agree that backing up static files sound unnecessary, but I don't think the added complexity of handling it is worth

I just pushed a new commit that only renames the variables, but the feature is untouched

thomasf commented:
> I like the idea to remove the flag settings and rely on the path, but that raises the question of what to do when the bucket is not set, should we raise an error or default to something else?

Either raise an exception (the ImproperlyConfigured one or whatever it's called) or default to the same bucket as normal files.. Decide what you think is best.

> I totally agree that backing up static files sound unnecessary, but I don't think the added complexity of handling it is worth..

You just need to only handle it here to implement it specifically for the media storage class.
https://github.com/py-pa/django-minio-storage/blob/master/minio_storage/storage.py#L350

I have several applications which uses a bunch of different custom application local sub classes of MinoStorage configured slightly different depending on use case.. I don't want django settings to affect custom subclasses, that's at least how it's designed to work right now.

rafaelsierra commented:
Can I keep the variable names and just move the logic to the media class?

thomasf commented:
I think that having the MEDIA infix somewhere makes sense, all other options which are specific to media or static have those words in the settings names if I'm not mistaken.
Just use the same pattern already in use for forwarding and/or overriding constructor arguments to the base class.

Commit:
Updated backup on delete to apply only to MinioMediaStorage

The settings available now applies only to MinioMediaStorage, but it is
still possible to other subclasses from MinioStorage to setup backup as
well, as long as the apropriate variables are assined during
initialisation

rafaelsierra commented:
Now the settings `MINIO_STORAGE_MEDIA_BACKUP_BUCKET` and `MINIO_STORAGE_MEDIA_BACKUP_FORMAT` applies only to the `MinioMediaStorage`, but the actual backup code is still on the `MinioStorage` so the feature can be used on other subclasses by setting `backup_format` and `backup_bucket` during initialization

The flag settings to enable/disable the feature has been dropped

thomasf commented:
If we need to expand the FORMAT later we can allow for it to be a function but let's not do that unless someone really needs it.

thomasf commented:
Please fix the documentation setting names, I will merge after that.

Commit:
Fixed variable name on docs

rafaelsierra commented:
I've finally fixed the variable names, sorry for the huge delay
